### ISSUE TITLE
[RFC] Recoverable parallel child process when system errors

### ISSUE THREAD

**Issue by yguedidi** [rectorphp/rector#7915]
TL;DR: an alternative/complement to [Allow to use a dedicated process for each job](https://github.com/rectorphp/rector-src/pull/3563)

While working on the unrelated [Allow to fully process files](https://github.com/rectorphp/rector-src/pull/3724) (more precisely [this commit](https://github.com/rectorphp/rector-src/pull/3724/commits/180577a847e132a00f2c56065a625fe3445d60ce)) and trying to [make PHPStan happy](https://github.com/rectorphp/rector-src/actions/runs/4855930240/jobs/8654964805?pr=3724) by reducing complexity, I found the following peace of code:

https://github.com/rectorphp/rector-src/blob/main/packages/Parallel/Application/ParallelFileProcessor.php#L178-L182

I was curious about why Rector quit all processes when a threshold of errors is reached?

Questioning around me, a reason that raised is to fail fast because we know the run will exit with a failure code anyway, preserving some resources and providing early feedback.
Are there other reasons?
I think this behavior is perfect when using dry run mode, but when not dry run, we may expect that Rector continue to refactor other files.

Then come the problem with a failure because of memory usage exceeded by a child process, the error that is often raised when running Rector on a big project.
And then a potential solution to that problem: have recoverable memory exceeded errors by starting a new child process that will replace the just killed one and make use of the newly available memory!

Then to avoid repeated errors because of a job that on its own uses a lot of memory and always fails, we could be smart and split the failing job in smaller jobs!
A safe way to split would be create a new job per file in the failing job (so job size = 1).
This way, an initial job of 20 files with 5 small files first, then 1 very big and followed by 14 other smaller ones, and that fails on the big file because of memory limit, Rector will have a chance to both process the big file and the other 14!
And if a job of 1 file fails, don't recover from that error and keep it as an error.

Compared to https://github.com/rectorphp/rector-src/pull/3563, as this will create new processes only on errors, it should not have the execution time impact the former one has.

@TomasVotruba @samsonasik what do you think? Would this be an acceptable solution?
I'll wait for an agreement before working on a PR here. :)

By implementing only this proposal, Rector could still use the whole available memory of the system, but at least it will run until the end.
After this, https://github.com/rectorphp/rector-src/pull/3563 will become a memory saving feature only, to not use the whole system memory, at the cost of execution time. Maybe still worth having for some use cases, like having memory efficient alternative to the time efficient default.

**Comment by staabm** [comment]
In my understanding a hard failure means that some basic problems appear (like unable to find certain classes via autoloading etc) and such problems can easily lead to rector take very wrong assumptions and transform code into garbage. 

I guess thats why we stop after a certain amount of errors since making the user wait a rather long time to finally get a unexpected result is not good DX

**Comment by yguedidi** [comment]
Good point!
Of course there is no intent to remove this threshold to stop early, but more at least for memory exceeded system errors to have a way to recover from them and retry with one file.

**Comment by TomasVotruba** [comment]
I agree with @staabm this should hard fail as fast as possible, as we don't want to make these errors happen.

**Comment by yguedidi** [comment]
@TomasVotruba maybe it wasn't clear in the proposal, but the fast failling will be preserved for real errors.
Idea here is specific to memory exceeded errors that could be recovered.
